src/analysis/retail/shaman/elemental/modules/talents/Stormkeeper.tsx

```diff
@@ -1,37 +1,264 @@
+ * before pressing SK
+ */
+const BASE_MAELSTROM_REQUIRED = 90;
+/** Maelstrom required on first WINDOW_START_SPELLS if SK was cast before pull. */
+const BASE_MAELSTROM_REQUIRED_PREPULL_CAST = 29;
+
+/** How much maelstrom the first spell in the SK damage rotation generates,
+ * presuming they cast the entire rotation after pressing SK. This value is
+ * used to determine if the user overcapped maelstrom on the start of the SK
+ * window.
+ */
+const FIRST_CAST_MAELSTROM_GENERATION = 12; // LvB
+
+class Stormkeeper extends MajorCooldown<SKCast> {
+  static dependencies = {
+    ...MajorCooldown.dependencies,
+    maelstromTracker: MaelstromTracker,
+    enemies: Enemies,
+    spellUsable: SpellUsable,
+    spellMaelstromCost: SpellMaelstromCost,
+  };
+
+  protected maelstromTracker!: MaelstromTracker;
+  protected enemies!: Enemies;
+  protected spellUsable!: SpellUsable;
+  protected abilities!: Abilities;
+  protected spellMaelstromCost!: SpellMaelstromCost;
+
+  activeWindow: SKCast | null = null;
+  lastSKHardcast: CastEvent | null = null;
+
+  stSpender: Talent;
+
   damageDoneByBuffedCasts = 0;
 
   constructor(options: Options) {
-    super(options);
+    super({ spell: TALENTS.STORMKEEPER_1_ELEMENTAL_TALENT }, options);
+
+    this.stSpender = this.selectedCombatant.hasTalent(TALENTS.ELEMENTAL_BLAST_TALENT)
+      ? TALENTS.ELEMENTAL_BLAST_TALENT
+      : TALENTS.EARTH_SHOCK_TALENT;
     this.active =
       this.selectedCombatant.getRepeatedTalentCount(TALENTS.STORMKEEPER_1_ELEMENTAL_TALENT) > 0;
     if (!this.active) {
       return;
     }
+
+    this.addEventListener(Events.cast.by(SELECTED_PLAYER), this.onEventDuringSK);
+    this.addEventListener(
+      Events.cast.by(SELECTED_PLAYER).spell(TALENTS.STORMKEEPER_1_ELEMENTAL_TALENT),
+      this.onSKCast,
+    );
     this.addEventListener(
-      Events.damage.by(SELECTED_PLAYER).spell(AFFECTED_ABILITIES),
+      Events.removebuff.by(SELECTED_PLAYER).spell(TALENTS.STORMKEEPER_1_ELEMENTAL_TALENT),
+      this.onSKFalloff,
+    );
+    this.addEventListener(
+      Events.damage.by(SELECTED_PLAYER).spell(SK_DAMAGE_AFFECTED_ABILITIES),
       this.onSKDamage,
     );
   }
 
+  onSKCast(cast: CastEvent) {
+    this.lastSKHardcast = cast;
+  }
+
+  onSKFalloff(event: RemoveBuffEvent) {
+    if (!this.activeWindow) {
+      return;
+    }
+
+    this.activeWindow.timeline.end = event.timestamp;
+
+    this.recordCooldown(this.activeWindow);
+    this.lastSKHardcast = null;
+    this.activeWindow = null;
+  }
+
+  onEventDuringSK(event: CastEvent) {
+    /* Don't include events that did not happen between hardcasts and the
+    SK buff falling off. */
+    if (!this.lastSKHardcast) {
+      return;
+    }
+
+    if (!this.activeWindow && WINDOW_START_SPELLS.includes(event.ability.guid)) {
+      this.activeWindow = {
+        event: this.lastSKHardcast,
+        maelstromOnCast: this.maelstromTracker.current,
+        flameshockDurationOnCast: this.enemies.getLongestDurationRemaining(
```

Comment:
Good point. This logic originates from when I stored these values on SK hardcast, but since I'm now storing on a damaging cast event, I have a target.